codegen/gce/src/expressions.rs

```diff
@@ -0,0 +1,364 @@
+                    )?);
+                    // create mapping (index in node graph: index in expressions vector)
+                    self.expressions_map.insert(index, expressions.len() - 1);
+                }
+                Operation::Sub(l, r) => {
+                    expressions.push(self.handle_transition_expression(
+                        ExpressionOperation::Sub,
+                        *l,
+                        *r,
+                    )?);
+                    self.expressions_map.insert(index, expressions.len() - 1);
+                }
+                Operation::Mul(l, r) => {
+                    expressions.push(self.handle_transition_expression(
+                        ExpressionOperation::Mul,
+                        *l,
+                        *r,
+                    )?);
+                    self.expressions_map.insert(index, expressions.len() - 1);
+                }
+                Operation::Exp(i, degree) => {
+                    match degree {
+                        0 => {
+                            // I decided that node^0 could be emulated using the product of 1*1, but perhaps there are better ways
+                            let index_of_1 = get_constant_index_by_value(1, self.constants)?;
+                            let const_1_node = NodeReference {
+                                node_type: NodeType::Const,
+                                index: index_of_1,
+                            };
+                            expressions.push(ExpressionJson {
+                                op: ExpressionOperation::Mul,
+                                lhs: const_1_node.clone(),
+                                rhs: const_1_node,
+                            });
+                        }
+                        1 => {
+                            let lhs = self.handle_node_reference(*i)?;
+                            let degree_index = get_constant_index_by_value(1, self.constants)?;
+                            let rhs = NodeReference {
+                                node_type: NodeType::Const,
+                                index: degree_index,
+                            };
+                            expressions.push(ExpressionJson {
+                                op: ExpressionOperation::Mul,
+                                lhs,
+                                rhs,
+                            });
+                        }
+                        _ => self.handle_exponentiation(&mut expressions, *i, *degree)?,
+                    }
+                    self.expressions_map.insert(index, expressions.len() - 1);
+                }
+                _ => {}
+            }
+        }
+        Ok(expressions)
+    }
+
+    /// Fills the `outputs` vector with indexes from `expressions` vector according to the `expressions_map`.
+    pub fn get_outputs(
+        &self,
+        expressions: &mut Vec<ExpressionJson>,
+    ) -> Result<Vec<usize>, ConstraintEvaluationError> {
```

Comment:
We can change the signature to the following to implement the builder pattern I suggested:

```
fn build_outputs(&self, ir: &AirIR) -> Result<Vec<usize>, ConstraintEvaluationError> {
```